File: core/src/main/java/com/delia/core/util/SharedPreUtil.java
package com.delia.core.util;

import android.content.Context;
import android.content.SharedPreferences;

import com.delia.core.BaseApplication;
import com.delia.core.exception.SharedInitialException;

import java.lang.reflect.Constructor;

/**
 * SharedPreferences工具类
 * @author xiong'MissDelia'zhengkun
 * 2020年7月17日14:36:57
 */
public class SharedPreUtil {

    private static final String TAG = "JMX_SHARED";

    // 创建一个写入器
    private SharedPreferences mPreferences;
    private static SharedPreUtil instance;

    public synchronized static SharedPreUtil getInstance() {
        if (instance == null) {
            instance = new SharedPreUtil();
        }
        return instance;
    }

    private SharedPreUtil() {
        mPreferences = BaseApplication.getApplication().getSharedPreferences(TAG, Context.MODE_PRIVATE);
    }

    /**
     * 存数据
     * @param key 关键字
     * @param value 值
     * @param <T> 只接受基本数据类型或者其包装类
     */
    public <T> void setValue(String key, T value) throws SharedInitialException {
        if (mPreferences == null) {
            throw new SharedInitialException("Fail to initialize SharedPreferences" +
                    ", SharedPreUtil is unavailable!");
        }
        String tempValue = String.valueOf(value);
        mPreferences.edit().putString(key, tempValue).apply();
    }

    /**
     * 取数据
     * @param key 关键字
     * @param def 默认数据
     * @param <T> 只接受基本数据类型或者其包装类
     */
    @SuppressWarnings("unchecked")
    public <T> T getValue(String key, T def) throws SharedInitialException {
        if (mPreferences == null) {
            throw new SharedInitialException("Fail to initialize SharedPreferences" +
                    ", SharedPreUtil is unavailable!");
        }
        try {
            String tempDef = String.valueOf(def);
            String tempValue = mPreferences.getString(key, tempDef);
            Class<T> clz = (Class<T>) def.getClass();
            Constructor<T> constructor = clz.getConstructor(String.class);
            return constructor.newInstance(tempValue);
        } catch (Exception e) {
            e.printStackTrace();
        }
        return def;
    }

    /**
     * 移除数据
     */
    public void removeKey(String key) {
        mPreferences.edit().remove(key).apply();
    }

}

File: core/src/main/java/cool/delia/core/utils/SharedPreUtil.kt
/*
 * 2016-2021 ©MissDelia 版权所有
 * "Anti 996" License Version 1.0
 */
package cool.delia.core.utils

import android.annotation.SuppressLint
import android.content.Context
import cool.delia.core.CoreApplication
import cool.delia.core.exception.CacheInitialException
import com.tencent.mmkv.MMKV

/**
 * MMKV的封装实现
 * @author xiong'MissDelia'zhengkun
 * @since V1.3.0
 * 2021/4/21 15:28
 */
object SharedPreUtil: MemoryCacheModel {

    /**
     * 全局持久缓存实例
     */
    private val cache = MMKV.defaultMMKV()

    /**
     * 从旧的Shared模型迁移数据
     */
    fun moveOldMemoryModel(memoryKey: String) {
        val oldShared = CoreApplication.getApplication().getSharedPreferences(memoryKey, Context.MODE_PRIVATE)
        if (cache != null) {
            cache.importFromSharedPreferences(oldShared)
            oldShared.edit().clear().apply()
        }
    }

    @SuppressLint("CommitPrefEdits")
    override fun <T> setValue(key: String, def: T) {
        if (cache == null) {
            throw CacheInitialException("Fail to initialize SharedPreferences" +
                    ", SharedPreUtil is unavailable!")
        }
        val tempValue: String = def.toString()
        cache.edit().putString(key, tempValue)
    }

    override fun <T> getValue(key: String, def: T): T {
        if (cache == null) {
            throw CacheInitialException("Fail to initialize SharedPreferences" +
                    ", SharedPreUtil is unavailable!")
        }
        try {
            val tempDef = def.toString()
            val tempValue: String? = cache.getString(key, tempDef)
            val clz = def!!::class.java
            val constructor = clz.getConstructor(String::class.java)
            return constructor.newInstance(tempValue)
        } catch (e: Exception) {
            e.printStackTrace()
        }
        return def
    }

    override fun hasKey(key: String): Boolean {
        if (cache == null) {
            throw CacheInitialException("Fail to initialize SharedPreferences" +
                    ", SharedPreUtil is unavailable!")
        }
        return cache.containsKey(key)
    }

    @SuppressLint("CommitPrefEdits")
    override fun removeKey(key: String) {
        if (cache == null) {
            throw CacheInitialException("Fail to initialize SharedPreferences" +
                    ", SharedPreUtil is unavailable!")
        }
        cache.edit().remove(key)
    }
}